opm/simulators/wells/MultisegmentWellEval.cpp

```diff
@@ -523,9 +523,6 @@ getResidualMeasureValue(const WellState& well_state,
         ++count;
     }
 
-    // if (count == 0), it should be converged.
```

Comment:
Maybe we discuss before. Can you explain the circumstances that this assert will fail? Maybe we can go through different approach avoid breaching the design of the function.

Reply:
Yes, when `iterateWellEqWithSwitching` is called for operability-update, it will do one iteration even though equations already are converged. I had some issues avoiding this, but looking at it now, I'm not sure why. Will re-check.

Reply:
It's actually a bit tricky to make sure `getResidualMeasureValue` is never called for converged equations. We might have a situation where the the equations are converged but not all constraints are satified. In this case `iterateWellEqWithSwitching` will continue until all constraints are met. If we skip calling  `getResidualMeasureValue` for one of the iterations, this will break the stagnation/oscillations-checking.

Reply:
Then from the code level without considering the exact scenario, there is something inconsistent between the `getWellConvergence` and the criteria used in the function `getResidualMeasureValue` including `getFiniteWellResiduals`. because if `getWellConvergence` is `true`, the function `getResidualMeasureValue` will not be reached.  It has been long time since the code got written, not totally sure whether they should be consistent. 

The following code can be related also, 
```c++
    // for stopped well, we do not enforce the following checking to avoid dealing with sign of near-zero values
    // for BHP or THP controlled wells, we need to make sure the flow direction is correct
    if (!baseif_.wellIsStopped() && baseif_.isPressureControlled(well_state)) {
        // checking the flow direction
        const double sign = baseif_.isProducer() ? -1. : 1.;
        const auto weight_total_flux = this->primary_variables_.getWQTotal() * sign;
        constexpr int dummy_phase = -1;
        if (weight_total_flux < 0.) {
            report.setWellFailed(
                    {CR::WellFailure::Type::WrongFlowDirection, CR::Severity::Normal, dummy_phase, baseif_.name()});
        }
    }
```


The above can be an isolated issue to be handled separately.

Reply:
Now I see, it is the following code that can break the assert. 

```c++
                if (switch_count > 0 && its_since_last_switch < min_its_after_switch) {
                    final_check = true;
                    its_since_last_switch = min_its_after_switch;
                } else {
                    break;
                }   
```
so it is good for now.